import "geometry.proto";
import "items.proto";

message Entity {
  extensions 100 to max;

  enum Direction {
    NORTH = 0;
    WEST = 1;
    SOUTH = 2;
    EAST = 3;
  }

  required string type = 1;
  required uint32 realm_id = 2;
  required Vector3 location = 3;
  required Direction direction = 4;
  required Rectangle bbox = 5;
}

message Drop {
  extend Entity {
    optional Drop ext = 104;
  }

  required Item item = 1;
}

message Building {
  extend Entity {
    optional Building ext = 105;
  }

  required Entity.Direction door_location = 1;
}

message Fixture {
  extend Entity {
    optional Fixture ext = 103;
  }

  required string fixture_type = 1;
}

message Actor {
  extend Entity {
    optional Actor ext = 101;
  }

  required string name = 1;
  required uint32 health = 2;
  required string gender = 3;
  required string body = 4;
  optional string facial = 5;
  optional string hair = 6;

  optional Item head_item = 7;
  optional Item torso_item = 8;
  optional Item legs_item = 9;
  optional Item feet_item = 10;
  optional Item weapon = 11;

  repeated Item inventory = 12;
}

message Player {
  extend Entity {
    optional Player ext = 106;
  }

  required bool online = 1;
}

message NPC {
  extend Entity {
    optional NPC ext = 107;
  }

  optional string behavior = 1;
}

message Teleporter {
  extend Entity {
    optional Teleporter ext = 108;
  }

  required uint32 teleport_realm_id = 1;
  required Vector3 teleport_location = 2;
}

message Tree {
  extend Entity {
    optional Tree ext = 109;
  }

  required string species = 1;

  enum GrowthStage {
    SEEDLING = 0;
    SAPLING = 1;
    MATURE = 2;
  }

  required GrowthStage growth_stage = 2;
}
